File: data-raw/check-checkpoint-sites.py
#!/usr/bin/env python3
"""M120 — every declared resume site still routes through the checkpoint guard.

What this checks, and what it deliberately does not:

  * IT CHECKS the sites declared in data-raw/checkpoint-sites.tsv. That file is
    an enumeration, not a discovery procedure, and this checker inherits exactly
    that bound. It cannot tell you a sixth harness has started resuming from a
    cache, and it does not claim to.
  * WHAT DOES cover the open-ended case is the run-time trace in
    checkpoint-guard.R: it watches deserialization as it happens, so a
    checkpoint read fails the run whatever spelling it was written in. This
    checker exists because that trace only fires when a harness actually runs,
    and the oracle harnesses need brms/Stan and hours of refits.

So: the trace is the guard, this is the regression net over the sites we know
about. Stdlib-only and R-free, to stay in the existing check-references job.

Usage:
  python3 data-raw/check-checkpoint-sites.py
  python3 data-raw/check-checkpoint-sites.py --self-test
"""

import re
import sys
import pathlib
import tempfile

ROOT = pathlib.Path(__file__).resolve().parent.parent
SITES = ROOT / "data-raw" / "checkpoint-sites.tsv"
GUARD = "data-raw/checkpoint-guard.R"


def read_sites(path=SITES):
    rows = []
    header = None
    for line in path.read_text().splitlines():
        if not line.strip() or line.startswith("#"):
            continue
        parts = line.split("\t")
        if header is None:
            header = parts
            continue
        rows.append(dict(zip(header, parts)))
    return rows


def strip_comments(text):
    """Drop R comments, leaving everything else (including line structure) put.

    Without this the checker reads commented-out code as live code, and
    `# source("data-raw/checkpoint-guard.R")` passes every routing check while
    the guard is switched off — which is exactly how a guard gets disabled
    during debugging and left that way. A `#` inside a string literal is not a
    comment, so quoting state is tracked rather than the line simply split on
    the first hash.
    """
    out = []
    for line in text.splitlines():
        quote = None
        i = 0
        while i < len(line):
            ch = line[i]
            if quote:
                if ch == "\\":
                    i += 2
                    continue
                if ch == quote:
                    quote = None
            elif ch in "\"'":
                quote = ch
            elif ch == "#":
                break
            i += 1
        out.append(line[:i] if i < len(line) else line)
    return "\n".join(out)


def check_site(site, root=ROOT):
    """Return a list of failure strings for one declared site."""
    out = []
    path = root / site["script"]
    if not path.exists():
        return [f"{site['script']}: declared in checkpoint-sites.tsv but absent"]
    text = strip_comments(path.read_text())

    if f'source("{GUARD}")' not in text:
        out.append(f"{site['script']}: does not source {GUARD}")

    # The resume read must go through the guard. A bare readRDS of the
    # checkpoint variable is the exact call this milestone replaced.
    if re.search(r"(?<![\w.$])readRDS\s*\(\s*ckpt\b", text):
        out.append(
            f"{site['script']}: reads its checkpoint with a bare readRDS(ckpt) "
            f"— route it through ckpt_read()"
        )

    # EVERY declared guard call, not any one of them. A store site that keeps
    # its file-level load and drops the per-entry ckpt_store_get() has stopped
    # checking the thing that actually goes stale, and an "any of these" test
    # stays silent through exactly that edit.
    required = [c for c in site.get("api", "").split(",") if c]
    if not required:
        out.append(f"{site['script']}: declares no guard calls")
    for call in required:
        if not re.search(rf"(?<![\w.$]){re.escape(call)}\s*\(", text):
            out.append(f"{site['script']}: never calls {call}()")
    if "ckpt_trace_assert()" not in text:
        out.append(
            f"{site['script']}: never calls ckpt_trace_assert() before writing output"
        )
    if "ckpt_trace_register(" not in text:
        out.append(f"{site['script']}: never registers its checkpoint with the trace")

    # Every declared parameter must actually appear in a spec call, or the
    # declaration and the code have drifted apart.
    declared = [p for p in site["params"].split(",") if p]
    if not declared:
        out.append(f"{site['script']}: declares no parameters")
    spec_blocks = re.findall(r"ckpt_spec\s*\((.*?)\n  \)", text, flags=re.S)
    spec_text = "\n".join(spec_blocks)
    for p in declared:
        if not re.search(rf"(?<![\w.]){re.escape(p)}\s*=", spec_text):
            out.append(
                f"{site['script']}: declared parameter '{p}' appears in no ckpt_spec() call"
            )
    for fn in (f for f in site["block"].split(",") if f):
        if f'"{fn}"' not in spec_text:
            out.append(
                f"{site['script']}: declared block function '{fn}' is in no ckpt_spec() call"
            )
    return out


def run_check(root=ROOT, sites=None):
    sites = sites if sites is not None else read_sites()
    failures = []
    for site in sites:
        failures.extend(check_site(site, root=root))
    return sites, failures


def mutations_for(site, src):
    """(label, mutated source) pairs for one site.

    Every declared guard call gets its own reversion, because dropping one of
    them is a real edit that leaves the others in place: on the store sites the
    per-entry check is the one that can vanish while the file-level load still
    looks like routing.
    """
    yield (
        "commented out the source() of the guard",
        src.replace(f'source("{GUARD}")', f'# source("{GUARD}")', 1),
    )
    yield (
        "commented out the pre-write trace assertion",
        src.replace("ckpt_trace_assert()", "# ckpt_trace_assert()", 1),
    )
    for call in (c for c in site.get("api", "").split(",") if c):
        yield (
            f"dropped every {call}() call",
            re.sub(rf"(?<![\w.$]){re.escape(call)}\s*\(", "bypassed_(", src),
        )


def self_test():
    """Mutation harness: revert a guard call in a copy of a real site and
    require this checker to notice. A checker that passes on a reverted guard
    is false coverage, which is the trap this repo keeps re-finding.

    Every declared site is probed, not a representative one: the sites do not
    share a guard API, so a harness that only ever mutates the first site says
    nothing about the four that route their reads a different way."""
    ok = True
    for site in read_sites():
        src = (ROOT / site["script"]).read_text()
        for label, mutated in mutations_for(site, src):
            if mutated == src:
                print(
                    f"FAIL self-test [{site['script']}]: cannot plant '{label}'"
                    " — the mutation changed nothing"
                )
                ok = False
                continue
            with tempfile.TemporaryDirectory() as td:
                fake_root = pathlib.Path(td)
                dest = fake_root / site["script"]
                dest.parent.mkdir(parents=True, exist_ok=True)
                dest.write_text(mutated)
                found = check_site(site, root=fake_root)
            if found:
                print(f"PASS self-test [{site['script']}]: {label} -> detected")
            else:
                print(f"FAIL self-test [{site['script']}]: {label} was NOT detected")
                ok = False

        # And the unmutated file must pass, or the checker reports on everything.
        if check_site(site):
            print(f"FAIL self-test [{site['script']}]: the unmutated site does not pass")
            ok = False
        else:
            print(f"PASS self-test [{site['script']}]: the unmutated site passes")
    return ok


def main(argv):
    if "--self-test" in argv:
        return 0 if self_test() else 1
    sites, failures = run_check()
    if failures:
        print(f"FAIL: {len(failures)} checkpoint-routing failure(s):")
        for f in failures:
            print("  -", f)
        return 1
    print(
        f"OK: {len(sites)} declared resume site(s) route through the checkpoint guard."
    )
    print(
        "(Declared sites only — the run-time trace in checkpoint-guard.R is what "
        "covers a site nobody declared.)"
    )
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: data-raw/check-checkpoint-sites.R
# M120 — every declared resume site still routes through the checkpoint guard.
#
# This replaces a text-matching checker that asked only whether a token appeared
# somewhere in the file. Three reversions walked past it: an assertion moved to
# after the fixture write, a registration handed a literal NULL, and a live
# per-entry read replaced by direct payload access while a dead copy of the call
# survived in an `if (FALSE)` block. None of those is visible to a search for
# text, so this reads the parsed source instead.
#
# What it checks, and what it deliberately does not:
#
#   * IT CHECKS the sites declared in data-raw/checkpoint-sites.tsv. That file is
#     an enumeration, not a discovery procedure, and this checker inherits
#     exactly that bound. It cannot tell you a sixth harness has started resuming
#     from a cache, and it does not claim to.
#   * WHAT DOES cover the open-ended case is the run-time trace in
#     checkpoint-guard.R, which installs itself when that file is sourced and
#     watches deserialization as it happens. This checker exists because that
#     trace only fires when a harness actually runs, and the four oracle
#     harnesses need brms/Stan and hours of refits.
#
# So: the trace is the guard, this is the regression net over the sites we know
# about.
#
#   Rscript data-raw/check-checkpoint-sites.R
#   Rscript data-raw/check-checkpoint-sites.R --self-test

guard_path <- "data-raw/checkpoint-guard.R"
sites_tsv <- "data-raw/checkpoint-sites.tsv"

# ---- reading the declaration -------------------------------------------------

read_sites <- function(path = sites_tsv) {
  lines <- readLines(path, warn = FALSE)
  lines <- lines[nzchar(trimws(lines)) & !startsWith(lines, "#")]
  parts <- strsplit(lines, "\t", fixed = TRUE)
  header <- parts[[1]]
  lapply(parts[-1], function(row) stats::setNames(as.list(row), header))
}

split_decl <- function(x) {
  out <- trimws(strsplit(x %||% "", ",", fixed = TRUE)[[1]])
  out[nzchar(out)]
}

`%||%` <- function(a, b) if (is.null(a)) b else a

# ---- walking the parsed source ----------------------------------------------
# Every check below is asked of the LIVE code. `if (FALSE) { ... }` is the
# cheapest way to keep a call's text while removing the call, so those branches
# are dropped before anything is counted; a dead call is not a call.

drop_dead <- function(e) {
  if (!is.call(e)) {
    return(e)
  }
  if (identical(e[[1]], as.name("if")) && identical(e[[2]], FALSE)) {
    return(if (length(e) >= 4L) drop_dead(e[[4]]) else NULL)
  }
  if (identical(e[[1]], as.name("if")) && identical(e[[2]], TRUE)) {
    return(drop_dead(e[[3]]))
  }
  as.call(lapply(as.list(e), function(x) if (is.null(x)) x else drop_dead(x)))
}

# Every call to `name` in the live source, as call objects.
calls_to <- function(exprs, name) {
  found <- list()
  walk <- function(e) {
    if (is.call(e)) {
      head <- e[[1]]
      # bare f(), pkg::f(), and pkg:::f() all count as a call to f
      nm <- if (is.name(head)) {
        as.character(head)
      } else if (
        is.call(head) &&
          is.name(head[[1]]) &&
          as.character(head[[1]]) %in% c("::", ":::")
      ) {
        as.character(head[[3]])
      } else {
        ""
      }
      if (identical(nm, name)) {
        found[[length(found) + 1L]] <<- e
      }
      # BY INDEX: an empty argument (the blank in `d[keep, , drop = FALSE]`)
      # binds a loop variable to the missing marker, and every later touch of
      # that variable raises "argument is missing". Indexing reads the same
      # object without binding it.
      args <- as.list(e)[-1]
      for (i in seq_along(args)) {
        if (!is.null(args[[i]]) && !identical(args[[i]], quote(expr = ))) {
          walk(args[[i]])
        }
      }
    }
  }
  for (e in exprs) {
    walk(e)
  }
  found
}

# The source line a call sits on, for the ordering check. Top-level scripts run
# in source order, so a line number is an execution order.
line_of <- function(exprs, name) {
  refs <- attr(exprs, "srcref")
  hits <- integer(0)
  for (i in seq_along(exprs)) {
    if (length(calls_to(exprs[i], name))) {
      hits <- c(hits, if (is.null(refs)) NA_integer_ else refs[[i]][1L])
    }
  }
  hits
}

parse_live <- function(path) {
  exprs <- parse(path, keep.source = TRUE)
  refs <- attr(exprs, "srcref")
  live <- lapply(exprs, drop_dead)
  live <- Filter(Negate(is.null), live)
  attr(live, "srcref") <- refs
  live
}

# ---- the checks --------------------------------------------------------------

check_site <- function(site, root = ".") {
  out <- character(0)
  path <- file.path(root, site$script)
  if (!file.exists(path)) {
    return(paste0(site$script, ": declared in ", sites_tsv, " but absent"))
  }
  exprs <- tryCatch(parse_live(path), error = function(e) NULL)
  if (is.null(exprs)) {
    return(paste0(site$script, ": does not parse"))
  }

  # 1. The guard is sourced, by a live call naming the guard file.
  sourced <- calls_to(exprs, "source")
  if (
    !any(vapply(
      sourced,
      function(e) {
        any(vapply(as.list(e)[-1], identical, logical(1), guard_path))
      },
      logical(1)
    ))
  ) {
    out <- c(out, paste0(site$script, ": no live source() of ", guard_path))
  }

  # 2. Every declared guard call is made, live. ALL of them: a site that keeps
  #    its file-level load and drops its per-entry read has stopped checking the
  #    thing that goes stale, and no single call stands in for the set.
  required <- split_decl(site$api)
  if (!length(required)) {
    out <- c(out, paste0(site$script, ": declares no guard calls"))
  }
  for (call_name in required) {
    if (!length(calls_to(exprs, call_name))) {
      out <- c(out, paste0(site$script, ": never calls ", call_name, "()"))
    }
  }

  # 3. The checkpoint is registered with the trace, with a real argument. A
  #    literal NULL registers nothing and leaves every later read unwatched.
  reg <- calls_to(exprs, "ckpt_trace_register")
  if (!length(reg)) {
    out <- c(out, paste0(site$script, ": never registers its checkpoint"))
  } else if (
    !any(vapply(
      reg,
      function(e) {
        length(e) >= 2L && !is.null(e[[2]]) && !identical(e[[2]], NULL)
      },
      logical(1)
    ))
  ) {
    out <- c(
      out,
      paste0(site$script, ": ckpt_trace_register() is given no real path")
    )
  }

  # 4. The pre-write assertion runs BEFORE the site's output write. Asserting
  #    after the fixture has been written is not a guard, it is a postmortem.
  assert_lines <- line_of(exprs, "ckpt_trace_assert")
  write_lines <- line_of(exprs, "saveRDS")
  if (!length(assert_lines)) {
    out <- c(out, paste0(site$script, ": never calls ckpt_trace_assert()"))
  } else if (length(write_lines) && min(assert_lines) > min(write_lines)) {
    out <- c(
      out,
      paste0(
        site$script,
        ": ckpt_trace_assert() (line ",
        min(assert_lines),
        ") comes after the first output write (line ",
        min(write_lines),
        ")"
      )
    )
  }

  # 5. The declared parameters appear in a ckpt_spec() call, in the declared
  #    ORDER — the mismatch message names the earliest differing one, so the
  #    order is part of the contract, not a formatting detail.
  specs <- calls_to(exprs, "ckpt_spec")
  if (!length(specs)) {
    out <- c(out, paste0(site$script, ": never builds a ckpt_spec()"))
  } else {
    declared <- split_decl(site$params)
    seen_any <- FALSE
    for (e in specs) {
      params <- e[["params"]]
      if (is.null(params) || !is.call(params)) {
        next
      }
      nms <- names(as.list(params))[-1]
      if (is.null(nms)) {
        next
      }
      seen_any <- TRUE
      if (!identical(nms, declared)) {
        out <- c(
          out,
          paste0(
            site$script,
            ": ckpt_spec() parameters (",
            paste(nms, collapse = ", "),
            ") do not match the declared set in declared order (",
            paste(declared, collapse = ", "),
            ")"
          )
        )
      }
    }
    if (!seen_any) {
      out <- c(out, paste0(site$script, ": no ckpt_spec() declares params"))
    }
  }

  # 6. The declared block roots and values are the ones the site passes.
  for (nm in c(split_decl(site$roots), split_decl(site$values))) {
    if (
      !any(vapply(
        specs,
        function(e) nm %in% all_strings(e),
        logical(1)
      ))
    ) {
      out <- c(
        out,
        paste0(
          site$script,
          ": declared block entry '",
          nm,
          "' is in no ckpt_spec()"
        )
      )
    }
  }
  out
}

all_strings <- function(e) {
  out <- character(0)
  walk <- function(x) {
    if (is.character(x)) {
      out <<- c(out, x)
    } else if (is.call(x) || is.list(x)) {
      ys <- as.list(x)
      for (i in seq_along(ys)) {
        if (!is.null(ys[[i]]) && !identical(ys[[i]], quote(expr = ))) {
          walk(ys[[i]])
        }
      }
    }
  }
  walk(e)
  out
}

run_check <- function(root = ".") {
  sites <- read_sites(file.path(root, sites_tsv))
  failures <- unlist(lapply(sites, check_site, root = root))
  list(sites = sites, failures = failures %||% character(0))
}

# ---- mutation self-test ------------------------------------------------------
# A checker that passes on a reverted guard is false coverage, which is the trap
# this repo keeps re-finding — and which this checker's predecessor fell into
# three more times. Every declared site is probed, not a representative one.

mutations_for <- function(site, src) {
  muts <- list(
    list(
      label = "commented out the source() of the guard",
      text = sub(
        paste0('source("', guard_path, '")'),
        paste0('# source("', guard_path, '")'),
        src,
        fixed = TRUE
      )
    ),
    list(
      label = "moved ckpt_trace_assert() after the output write",
      text = paste0(
        sub("ckpt_trace_assert()", "invisible(TRUE)", src, fixed = TRUE),
        "\nckpt_trace_assert()\n"
      )
    ),
    list(
      label = "registered a literal NULL instead of the checkpoint path",
      text = sub(
        "ckpt_trace_register(",
        "ckpt_trace_register(NULL) # ",
        src,
        fixed = TRUE
      )
    )
  )
  for (call_name in split_decl(site$api)) {
    # Drop every live call, but keep a DEAD copy: this is the reversion that
    # walked past a text search, and the one a debugging edit actually leaves.
    muts[[length(muts) + 1L]] <- list(
      label = paste0("replaced every live ", call_name, "() with a dead copy"),
      text = paste0(
        gsub(
          paste0("(?<![\\w.$])", call_name, "\\s*\\("),
          "bypassed_(",
          src,
          perl = TRUE
        ),
        "\nif (FALSE) {\n  ",
        call_name,
        "(a, b, c)\n}\n"
      )
    )
  }
  muts
}

self_test <- function() {
  ok <- TRUE
  for (site in read_sites()) {
    src <- paste(readLines(site$script, warn = FALSE), collapse = "\n")
    for (m in mutations_for(site, src)) {
      if (identical(m$text, src)) {
        cat(
          "FAIL self-test [",
          site$script,
          "]: cannot plant '",
          m$label,
          "' -- the mutation changed nothing\n",
          sep = ""
        )
        ok <- FALSE
        next
      }
      td <- tempfile("ckpt-selftest-")
      dir.create(file.path(td, dirname(site$script)), recursive = TRUE)
      writeLines(m$text, file.path(td, site$script))
      file.copy(sites_tsv, file.path(td, sites_tsv))
      found <- check_site(site, root = td)
      unlink(td, recursive = TRUE)
      if (length(found)) {
        cat(
          "PASS self-test [",
          site$script,
          "]: ",
          m$label,
          " -> detected\n",
          sep = ""
        )
      } else {
        cat(
          "FAIL self-test [",
          site$script,
          "]: ",
          m$label,
          " was NOT detected\n",
          sep = ""
        )
        ok <- FALSE
      }
    }
    if (length(check_site(site))) {
      cat(
        "FAIL self-test [",
        site$script,
        "]: the unmutated site does not pass\n",
        sep = ""
      )
      ok <- FALSE
    } else {
      cat(
        "PASS self-test [",
        site$script,
        "]: the unmutated site passes\n",
        sep = ""
      )
    }
  }
  ok
}

# ---- entry point -------------------------------------------------------------

if (sys.nframe() == 0L) {
  args <- commandArgs(trailingOnly = TRUE)
  if ("--self-test" %in% args) {
    quit(status = if (self_test()) 0L else 1L)
  }
  res <- run_check()
  if (length(res$failures)) {
    cat(
      "FAIL: ",
      length(res$failures),
      " checkpoint-routing failure(s):\n",
      sep = ""
    )
    for (f in res$failures) {
      cat("  - ", f, "\n", sep = "")
    }
    quit(status = 1L)
  }
  cat(
    "OK: ",
    length(res$sites),
    " declared resume site(s) route through the checkpoint guard.\n",
    sep = ""
  )
  cat(
    "(Declared sites only -- the run-time trace in checkpoint-guard.R is what",
    " covers a site nobody declared.)\n",
    sep = ""
  )
}
